File: python/realm_sigil/handler.py
"""HTTP handler helpers for Python web servers."""

import json
import os
import platform
import socket
import subprocess
import sys
import time
from datetime import datetime, timezone

from . import generate_name, version_dict

_start_time = time.time()
_start_iso = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def _git_info(project_dir: str | None = None) -> dict:
    """Gather git info from the working directory."""
    cwd = project_dir or os.getcwd()
    info = {"hash": "dev", "branch": "unknown", "dirty": False}

    try:
        info["hash"] = subprocess.run(
            ["git", "rev-parse", "--short", "HEAD"],
            capture_output=True, text=True, cwd=cwd
        ).stdout.strip() or "dev"

        info["branch"] = subprocess.run(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            capture_output=True, text=True, cwd=cwd
        ).stdout.strip() or "unknown"

        result = subprocess.run(
            ["git", "diff", "--quiet"],
            capture_output=True, cwd=cwd
        )
        info["dirty"] = result.returncode != 0
    except FileNotFoundError:
        pass

    return info


def make_version_response(
    name: str,
    description: str,
    realm: str,
    repo: str,
    project_dir: str | None = None,
) -> dict:
    """Build a complete version response with live system info."""
    git = _git_info(project_dir)

    return version_dict(
        name=name,
        description=description,
        realm=realm,
        repo=repo,
        hash=git["hash"],
        branch=git["branch"],
        dirty=git["dirty"],
        built=_start_iso,  # For Python, built ≈ started (no compilation)
        started=_start_iso,
        uptime=int(time.time() - _start_time),
        runtime=f"python{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
        os_info=f"{sys.platform}/{platform.machine()}",
        host=socket.gethostname(),
        pid=os.getpid(),
    )


def version_handler(
    name: str,
    description: str,
    realm: str,
    repo: str,
    project_dir: str | None = None,
):
    """Return a handler function suitable for Python HTTP servers.

    Works with http.server.BaseHTTPRequestHandler-style servers.
    Returns a callable that produces (status_code, headers, body_bytes).
    """
    def handle():
        body = json.dumps(
            make_version_response(name, description, realm, repo, project_dir),
            indent=2,
        ).encode()
        headers = {
            "Content-Type": "application/json",
            "Cache-Control": "no-cache",
            "Access-Control-Allow-Origin": "*",
        }
        return 200, headers, body

    return handle

File: js/handler.js
/**
 * HTTP handler helpers for Node.js servers (Express, Next.js, Vercel).
 */

const { execFileSync } = require('child_process');
const os = require('os');
const { versionObject } = require('./index');

const startTime = Date.now();
const startISO = new Date().toISOString();

function gitInfo(cwd) {
  const info = { hash: 'dev', branch: 'unknown', dirty: false };
  try {
    info.hash = execFileSync('git', ['rev-parse', '--short', 'HEAD'], { cwd, encoding: 'utf8' }).trim() || 'dev';
    info.branch = execFileSync('git', ['rev-parse', '--abbrev-ref', 'HEAD'], { cwd, encoding: 'utf8' }).trim() || 'unknown';
    try {
      execFileSync('git', ['diff', '--quiet'], { cwd });
      info.dirty = false;
    } catch {
      info.dirty = true;
    }
  } catch {
    // git not available
  }
  return info;
}

/**
 * Build a complete version response with live system info.
 */
function makeVersionResponse(name, description, realm, repo, cwd) {
  const git = gitInfo(cwd);
  return versionObject({
    name,
    description,
    realm,
    repo,
    hash: git.hash,
    branch: git.branch,
    dirty: git.dirty,
    built: startISO,
    started: startISO,
    uptime: Math.floor((Date.now() - startTime) / 1000),
    runtime: `node${process.version}`,
    os: `${process.platform}/${process.arch}`,
    host: os.hostname(),
    pid: process.pid,
  });
}

/**
 * Next.js API route handler.
 * Usage: export default nextHandler('myapp', 'My app', 'forge', 'https://github.com/jphein/myapp')
 */
function nextHandler(name, description, realm, repo, cwd) {
  return (req, res) => {
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.status(200).json(makeVersionResponse(name, description, realm, repo, cwd));
  };
}

/**
 * Vercel serverless handler.
 * Usage: export default vercelHandler('myapp', 'My app', 'tarot', 'https://github.com/jphein/myapp')
 */
function vercelHandler(name, description, realm, repo, cwd) {
  return (req, res) => {
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.status(200).json(makeVersionResponse(name, description, realm, repo, cwd));
  };
}

/**
 * Express middleware.
 * Usage: app.get('/api/version', expressHandler('myapp', 'My app', 'forge', 'https://...'))
 */
function expressHandler(name, description, realm, repo, cwd) {
  return (req, res) => {
    res.set('Cache-Control', 'no-cache');
    res.set('Access-Control-Allow-Origin', '*');
    res.json(makeVersionResponse(name, description, realm, repo, cwd));
  };
}

module.exports = { makeVersionResponse, nextHandler, vercelHandler, expressHandler, gitInfo };
